### DISCUSSION TITLE
Providing help for scripts

### DISCUSSION THREAD

**Discussion by ineu** [nushell/nushell#9168] [Q&A]
At the moment I use Just as a task runner. It provides very convenient syntax for tasks, including tasks descriptions and list of tasks. For example, given a Justfile named `dev` that looks like this:
```
#!/usr/bin/env -S just --justfile

@_default:
    just -l -f {{justfile()}}

# Run `rails console` on staging
console-stage:
    ssh ...
```
I can run `dev` to see the list of tasks and `dev console-stage` to run a specific task.

Most my tasks inside the Justfile are written in Nu, so I would like to remove dependency on `just` and use a script in Nu instead. And nu's `def`s behave pretty much the same, I can achieve both help messages and list of defs via the standard syntax, like this:
```
def dev [] {
  help dev
}


# Run `rails console` on staging
def "dev console-stage" [] {
  ...
}
```

The problem with this approach is: to achieve the same behavior I have to `import` this nu file. But I don't want to import it, I'd like to use it as a script. In such case the syntax changes to:
```
#!/bin/nu

def main [] {
  help main
}

# Run `rails console` on staging
def "main console-stage" [] {
  ...
}
```

But the behavior changes as well. While I can run the subcommand via `dev console-stage`, the help output now always contains "main":
```
Usage:
  > main

Subcommands:
  main console-stage - Run `rails console` on staging

Flags:
  -h, --help - Display the help message for this command
```
which is harmless (if you know it) but annoying.

Am I missing something? Should the scripts be used in another way?

**Comment by fdncred** [comment]
Just thinking out loud here but one possible solution is to use a hook like this
```
    env_change: {
      PWD: [
      { 
        condition: {|_, after| not ($after | path join 'toolkit.nu' | path exists)}
        code: "hide toolkit"
      },
      {
        condition: {|_, after| $after | path join 'toolkit.nu' | path exists}
        code: "
        print $'(ansi default_underline)(ansi default_bold)toolkit(ansi reset) module (ansi green_italic)detected(ansi reset)...'
        print $'(ansi yellow_italic)activating(ansi reset) (ansi default_underline)(ansi default_bold)toolkit(ansi reset) module with `(ansi default_dimmed)(ansi default_italic)use toolkit.nu(ansi reset)`'
        use toolkit.nu
        "
      }
      ]
    }
```
This is borrowed from @amtoine and tweaked a little. What happens here is when the PWD changes and a file named toolkit.nu exists, it's automatically used. So, then one can call any function out of the toolkit.nu module.

**Comment by ineu** [comment]
Yes, but still I don't want to import the code :)
Scripts are great to isolate the logic, so I'd like the script name to be the only entrypoint.

What I want to understand is if the `def main [] { help main }` approach is the proper one. If so, then it looks like a feature request to make `help` aware that it's running in the context of the script and should replace `main` with the script's name.

**Comment by fdncred** [comment]
We're aware of the problem with `help` not knowing the proper name when `main` is used like we're talking about. I'm sure there's an issue with that in there somewhere in our pile of issue. Feel free to find the issue and add your two cents.

**Reply by ineu** [reply]
Great, thank you
